### PR TITLE
DeepSpeed ZeRO Update

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Adds DeepSpeed ZeRO 3 to the DeepSpeed plugin. Due to the nature of the plugin, we've had to make a few changes summarized below.

- Plugin has to manage saving logic as DeepSpeed only supports non-elastic checkpointing; i.e if you train on 8 GPUs, you must load the model on 8 GPUs. This will be addressed over the coming weeks as PRs are merged into DeepSpeed
- We've introduced a sharded context manager to create layers and shard them instantly.
- DeepSpeed in ZeRO 3 has to wrap the model with optimizers even in inference. This is a limitation of DeepSpeed where the sharding logic is tied to the Optimizer, hence we only get proper sharding logic when an optimizer is passed through. Overall this means we have to allocate some memory to optimizer state (which is already done within training).

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: requirements/extra.txt [hunk 1]
```diff
@@ -9,3 +9,4 @@ onnxruntime>=1.3.0
 hydra-core>=1.0
 # todo: when switch to standard package stream, drop `fairscale` from hard mocked docs libs
 https://github.com/PyTorchLightning/fairscale/archive/pl_1.2.0.zip
+deepspeed
```
Comment (SeanNaren) [inline requirements/extra.txt:12]:
```suggestion
deepspeed==0.3.13
```

File: pytorch_lightning/utilities/imports.py [hunk 2]
```diff
@@ -63,14 +63,23 @@ def _compare_version(package: str, op, version) -> bool:
...
+
 _IS_WINDOWS = platform.system() == "Windows"
 _IS_INTERACTIVE = hasattr(sys, "ps1")  # https://stackoverflow.com/a/64523765
 _TORCH_LOWER_EQUAL_1_4 = _compare_version("torch", operator.le, "1.5.0")
 _TORCH_GREATER_EQUAL_1_6 = _compare_version("torch", operator.ge, "1.6.0")
 _TORCH_GREATER_EQUAL_1_7 = _compare_version("torch", operator.ge, "1.7.0")
 _TORCH_GREATER_EQUAL_1_8 = _compare_version("torch", operator.ge, "1.8.0")
-
-_KINETO_AVAILABLE = torch.profiler.kineto_available() if _TORCH_GREATER_EQUAL_1_8 else False
+_KINETO_AVAILABLE = _is_kineto_available()
```
Comment (SeanNaren) [inline pytorch_lightning/utilities/imports.py:82]: Why has this been changed in this PR? can we revert this?

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -771,7 +771,7 @@ def increment_accumulated_grad_global_step(self):
 
         # progress global step according to grads progress
         if num_accumulated_batches_reached or num_training_batches_reached:
-            self.trainer.global_step += 1
+            self.trainer.training_type_plugin.increment_accumulated_grad_global_step(self.trainer)
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/training_loop.py:774]: We should go through the accelerator

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -1086,10 +1086,14 @@ def call_setup_hook(self, model: LightningModule) -> None:
     def call_configure_sharded_model(self, model: LightningModule) -> None:
         # Call configure sharded model hook if accelerator requests. In some cases
         # we will not call the hook; the hook has initialized the sharded model for example.
-        if self.accelerator.call_configure_sharded_model_hook:
+
+        # used on the model if the user re-create a trainer with resume_from_checkpoint
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/trainer.py:1090]:
```suggestion
        # check the model if the user re-create a trainer with resume_from_checkpoint
```

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 5]
```diff
@@ -123,15 +120,15 @@ def restore_model_state(self, model: LightningModule, checkpoint) -> None:
         # restore model state_dict
         model.load_state_dict(checkpoint['state_dict'])
 
-    def restore_training_state(self, checkpoint):
+    def restore_training_state(self, checkpoint, load_optimizer_states: bool = True):
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:123]: Code conv is complaining that this `load_optimizer_states` function hasn't been tested. We probably want to add a test to ensure when this function is called and this variable is turned off via the training type plugin, we do not load the optimizer state.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 6]
```diff
@@ -197,6 +198,45 @@ def setup_optimizers_in_pre_dispatch(self) -> bool:
...
+        # hook: give user access to checkpoint if needed.
+        self.lightning_module.on_load_checkpoint(ckpt)
+        self.lightning_module.load_state_dict(ckpt['state_dict'])
+        return ckpt, True
+
+    def increment_accumulated_grad_global_step(self, trainer) -> None:
+        trainer.global_step += 1
+
+    @contextlib.contextmanager
+    def model_parallel_context(self) -> Generator:
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:230]: Merge has gone wrong here, this is already been merged into master as `model_sharded_context`. I think this should be removed, and we should refactor the API to use `model_sharded_context`

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 7]
```diff
@@ -232,28 +276,71 @@ def _initialize_deepspeed_train(self, model):
...
+            inference_config.update({"fp16": self.config["fp16"]})
+        if self.zero_stage_3:
+            inference_config.update({
+                "zero_allow_untested_optimizer": self.config['zero_allow_untested_optimizer'],
+                "zero_optimization": self.config['zero_optimization'],
+            })
+        # Remove all module hooks before initializing new model
+        remove_module_hooks(model)
+        model, _, _, _ = deepspeed.initialize(
+            args=SimpleNamespace(local_rank=self.local_rank),
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:336]: This should be removed as it's not necessary anymore I think

File: pytorch_lightning/utilities/imports.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/imports.py:82]: Yes, removed.

File: pytorch_lightning/trainer/training_loop.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:774]: Yes, done !

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:123]: Yes.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:230]: Done !

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/deepspeed.py:336]: Done !

File: requirements/extra.txt (hunk 1 above)
Comment (tchaton) [reply requirements/extra.txt:12]: Converted to `deepspeed>=0.3.13`

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 8]
```diff
@@ -232,28 +275,70 @@ def _initialize_deepspeed_train(self, model):
...
         )
+        self._set_deepspeed_activation_checkpointing()
 
         # set optimizer for save/load, but deepspeed manages the specific optimizer logic
         self.lightning_module.trainer.optimizers = [optimizer]
+        self.lightning_module.trainer.schedulers = [lr_scheduler]
         self.model = model
 
+    @contextlib.contextmanager
+    def model_sharded_context(self) -> Generator:
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/deepspeed.py:292]:
```suggestion
    def model_sharded_context(self) -> Generator[None, None, None]:
```

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 9]
```diff
@@ -333,8 +424,83 @@ def _format_precision_config(self):
...
+        return super().restore_model_state_from_ckpt_path(ckpt_path, map_location=map_location)
+
+    def _accumulated_batches_reached(self, total_batch_idx: int) -> bool:
+        return total_batch_idx % self._original_accumulate_grad_batches == 0
+
+    def compute_new_global_step(self, total_batch_idx: int, current_global_step: int) -> int:
+        if self._original_accumulate_grad_batches is None:
+            return current_global_step + 1
+        else:
+            if self._accumulated_batches_reached(total_batch_idx, ):
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/deepspeed.py:504]: Isn't this what should_accumulate in some other parts is for? This looks like duplication to me. Can we somehow reuse the accumulation checks we have somewhere else?

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/deepspeed.py:292]: Done !

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/deepspeed.py:504]: slightly different logic. But removed the function.

File: pytorch_lightning/trainer/training_loop.py [hunk 10]
```diff
@@ -771,7 +771,8 @@ def increment_accumulated_grad_global_step(self):
 
         # progress global step according to grads progress
         if num_accumulated_batches_reached or num_training_batches_reached:
-            self.trainer.global_step += 1
+            self.trainer.global_step = self.trainer.accelerator.compute_new_global_step(
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:774]: Change for `update_global_step`

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 11]
```diff
@@ -333,8 +424,83 @@ def _format_precision_config(self):
...
+            self.lightning_module.on_load_checkpoint(client_state)
+            return client_state, False
+        return super().restore_model_state_from_ckpt_path(ckpt_path, map_location=map_location)
+
+    def _accumulated_batches_reached(self, total_batch_idx: int) -> bool:
+        return total_batch_idx % self._original_accumulate_grad_batches == 0
+
+    def compute_new_global_step(self, total_batch_idx: int, current_global_step: int) -> int:
+        if self._original_accumulate_grad_batches is None:
+            return current_global_step + 1
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/deepspeed.py:502]: Yes. Done !

File: requirements/extra.txt [hunk 12]
```diff
@@ -9,3 +9,4 @@ onnxruntime>=1.3.0
 hydra-core>=1.0
 # todo: when switch to standard package stream, drop `fairscale` from hard mocked docs libs
 https://github.com/PyTorchLightning/fairscale/archive/pl_1.2.0.zip
+deepspeed==0.3.13
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply requirements/extra.txt]: woops, @awaelchli is right, definitely want this to be >=

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 13]
```diff
@@ -144,6 +164,16 @@ def __init__(
 
             min_loss_scale: The minimum FP16 dynamic loss scaling value (Default: 1000)
 
+            partition_activations: Enables partition activation when used with ZeRO stage 3.
+                Still requires you to wrap your forward functions in deepspeed.checkpointing.checkpoint.
+                See https://www.deepspeed.ai/tutorials/megatron/#deepspeed-activation-checkpoints-optional
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/deepspeed.py:169]:
```suggestion
                See `deepspeed tutorial <https://www.deepspeed.ai/tutorials/megatron/#deepspeed-activation-checkpoints-optional>`_
```

Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/deepspeed.py:169]: for marking it as a clickable link

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 14]
```diff
@@ -144,6 +164,16 @@ def __init__(
...
+
+            cpu_checkpointing: Offloads partitioned activations to CPU if ``partition_activations`` is enabled
+
+            contiguous_memory_optimization: Copies partitioned activations so that they are contiguous in memory.
+                Not supported by all models
+
+            synchronize_checkpoint_boundary: Insert ``torch.cuda.synchronize()`` at each checkpoint boundary.
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/deepspeed.py:176]:
```suggestion
            synchronize_checkpoint_boundary: Insert :func:`torch.cuda.synchronize` at each checkpoint boundary.
```
for cross referencing

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 15]
```diff
@@ -37,6 +37,16 @@
     import deepspeed
 
 
+def remove_module_hooks(model: torch.nn.Module) -> None:
+    for module in model.modules():
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:41]:
```suggestion
    # TODO: awaiting this feature to move upstream to DeepSpeed
    for module in model.modules():
```

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 16]
```diff
@@ -232,28 +276,70 @@ def _initialize_deepspeed_train(self, model):
...
+        # todo: Currently DeepSpeed requires optimizers at inference to partition weights correctly
+        optimizer, lightning_scheduler, optimizer_frequencies = None, None, None
+        if "optimizer" not in self.config:
+            rank_zero_info(
+                "You have not specified an optimizer or scheduler within the DeepSpeed config."
+                "Using `configure_optimizers` to define optimizer and scheduler."
+            )
+            optimizer, lightning_scheduler, optimizer_frequencies = self._init_scheduler_optimizer()
+        inference_config = {
+            # todo: this is required for DeepSpeed throughput timers
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:323]:
```suggestion
            # todo: this is required for DeepSpeed throughput timers, so deepspeed throughput timers will be incorrect
```

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 17]
```diff
@@ -282,6 +368,13 @@ def optimizer_step(self, optimizer: torch.optim.Optimizer, lambda_closure: Calla
         # internally, the engine has a reference to the optimizer already.
         self.model.step(**kwargs)
 
+    def _handle_gradient_accumulation_steps(self):
+        if self.config.get("gradient_accumulation_steps") > 1:
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:372]: @tchaton could we add some comments about what's happening here and why?

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 16 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/deepspeed.py]: Done !

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 17 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/deepspeed.py:373]: Done !

**Review by SeanNaren — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 18]
```diff
@@ -197,6 +198,45 @@ def setup_optimizers_in_pre_dispatch(self) -> bool:
         """
         return False
 
+    def restore_model_state_from_ckpt_path(
+        self,
+        ckpt_path: str,
+        map_location: Callable = lambda storage, loc: storage,
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:204]: what does this lambda does, return input arg or None?

Comment (Borda) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:204]: so is it `lambda x: x`
or `lambda x: None`

**Review by Borda — APPROVED** [review]
